### PR TITLE
#163682079 Implement login controller

### PR DESCRIPTION
**Pull request by bobbylemm**
#### What does this PR do?
Create a controller for user login
#### Description of Task to be completed?
Have the following endpoints working

`post /api/v1/login`

#### How should this be manually tested?

After cloning the repo, cd into it and RUN `npm install` and then `npm test`

- Using postman, go to `localhost:3000/api/v1/login`
- Provide a json object with the following fields
`email`
`password`

#### What are the relevant pivotal tracker stories?
#163682079(https://www.pivotaltracker.com/story/show/163682079)

### CODE REVIEW

**Review by omenkish — DISMISSED** [review]

File: server/controllers/userController.js [hunk 1]
```diff
@@ -23,35 +25,81 @@ class UserController {
...
+      const hashed = await baseHelpers.hashPassword(password);
+      await Users.findOrCreate({
+        where: {
+          [Op.or]: [{ email }, { username }]
+        },
         defaults: {
+          email,
           username,
-          password
+          password: hashed
```
Comment (p-funky) [inline server/controllers/userController.js]: There is nothing wrong with this but I am curious as to whether you are aware that you can do a before create and before update hook on your user model for the password to be hashed.

Comment (bobbylemm) [reply server/controllers/userController.js]: I recently became aware of this, I wanted to handle this in the controller

Comment (p-funky) [reply server/controllers/userController.js]: IMO though, I believe it's safer to use the hooks as once they are there, you don't need to try to remember whether you are hashing or not. With this, you'll still need to add similar code to your resetPassword method (and possibly your update user method).

**Review by omenkish — DISMISSED** [review]

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by primuse — DISMISSED** [review]

File: server/controllers/userController.js [hunk 2]
```diff
@@ -23,35 +23,95 @@ class UserController {
...
-        role: response.role
-      };
-      const token = assignToken(payload);
-      return res
-        .header('x-auth-token', token)
-        .status(201)
-        .json({
-          message: 'successfully registered to authors haven'
-        });
+      );
```
[the remark this answers is not in the document]
Comment (p-funky) [reply server/controllers/userController.js:66]: Seems houndci does not fancy prettifying. Ignoring and moving on please.

File: server/controllers/userController.js [hunk 3]
```diff
@@ -23,35 +23,95 @@ class UserController {
...
+        },
         defaults: {
+          email,
           username,
           password
         }
-      }).spread((user, created) => {
-        if (!created) {
-          return errorMessage(res, 409, 'this email or username already exists');
+      }).spread(
```
[the remark this answers is not in the document]
Comment (p-funky) [reply server/controllers/userController.js:35]: Seems houndci does not fancy prettifying. Ignoring and moving on please.

**Review by p-funky — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -28,6 +28,7 @@
 		"@sendgrid/mail": "^6.3.1",
 		"babel-core": "^7.0.0-bridge.0",
 		"babel-jest": "^24.0.0",
+		"bcrypt": "^3.0.3",
```
Comment (EfeAgare) [inline package.json:31]: Never mind that.

**Comment by EfeAgare** [comment]
You know what, never mind.

**Pull request merged**